crates/proof-of-sql/src/base/database/join_util.rs

```diff
@@ -1,19 +1,93 @@
 use super::{
     apply_column_to_indexes,
     order_by_util::{compare_indexes_by_columns, compare_single_row_of_tables},
-    Column, ColumnRepeatOp, ElementwiseRepeatOp, RepetitionOp, Table, TableOperationError,
-    TableOperationResult, TableOptions,
+    union_util::column_union,
+    Column, ColumnOperationResult, ColumnRepeatOp, ElementwiseRepeatOp, RepetitionOp, Table,
+    TableOperationError, TableOperationResult, TableOptions,
 };
 use crate::base::{
     map::{IndexMap, IndexSet},
     scalar::Scalar,
 };
-use alloc::vec::Vec;
+use alloc::{vec, vec::Vec};
 use bumpalo::Bump;
 use core::cmp::Ordering;
 use itertools::Itertools;
 use sqlparser::ast::Ident;
 
+/// Compute the set union of two slices of columns, deduplicate and sort the result.
+///
+/// Notes
+/// 1. This is mostly used for joins.
+/// 2. We do not check whether columns in the args have the same length, as we assume that the columns in an arg are already from the same table.
+pub(crate) fn ordered_set_union<'a, S: Scalar>(
+    left_on: &[Column<'a, S>],
+    right_on: &[Column<'a, S>],
+    alloc: &'a Bump,
+) -> TableOperationResult<Vec<Column<'a, S>>> {
+    //1. Union the columns
+    if left_on.len() != right_on.len() {
+        return Err(TableOperationError::JoinWithDifferentNumberOfColumns {
+            left_num_columns: left_on.len(),
+            right_num_columns: right_on.len(),
+        });
+    }
+    if left_on.is_empty() {
+        return Ok(Vec::new());
+    }
+    let raw_union = left_on
+        .iter()
+        .zip(right_on.iter())
+        .map(|(left, right)| column_union(&[left, right], alloc, left.column_type()))
+        .collect::<ColumnOperationResult<Vec<_>>>()?;
```

Comment:
It's possible (but more complex) to eliminate this `collect`. We can save it for later, but add some tracing to this method in case it's slow.